src/vcpkg/install.cpp

```diff
@@ -740,122 +760,165 @@ namespace vcpkg
                 }
             }
 
-            ret.header_only = is_header_only;
+            ret.header_only = !has_binaries && !header_path.empty();
 
-            if (library_targets.empty())
+            // Post-process cmake config data
+            bool has_targets_for_output = false;
+            for (auto&& package : config_packages)
             {
-                if (is_header_only && !header_path.empty())
+                const auto library_target_pair = library_targets.find(package.dir);
+                if (library_target_pair == library_targets.end()) continue;
+
+                auto& targets = library_target_pair->second;
+                if (!targets.empty())
                 {
-                    static auto cmakeify = [](std::string name) {
-                        auto n = Strings::ascii_to_uppercase(Strings::replace_all(std::move(name), "-", "_"));
-                        if (n.empty() || ParserBase::is_ascii_digit(n[0]))
-                        {
-                            n.insert(n.begin(), '_');
-                        }
-                        return n;
-                    };
+                    if (!package.name.empty()) has_targets_for_output = true;
 
-                    const auto name = cmakeify(bpgh.spec.name());
-                    auto msg = msg::format(msgHeaderOnlyUsage, msg::package_name = bpgh.spec.name()).extract_data();
-                    Strings::append(msg, "\n\n");
-                    Strings::append(msg, "    find_path(", name, "_INCLUDE_DIRS \"", header_path, "\")\n");
-                    Strings::append(msg, "    target_include_directories(main PRIVATE ${", name, "_INCLUDE_DIRS})\n\n");
+                    Util::sort_unique_erase(targets, [](const std::string& l, const std::string& r) {
+                        if (l.size() < r.size()) return true;
+                        if (l.size() > r.size()) return false;
+                        return l < r;
+                    });
 
-                    ret.message = std::move(msg);
+                    static const auto is_namespaced = [](const std::string& target) {
+                        return Strings::contains(target, "::");
+                    };
+                    if (Util::any_of(targets, is_namespaced))
+                    {
+                        Util::erase_remove_if(targets, [](const std::string& t) { return !is_namespaced(t); });
+                    }
                 }
+                ret.cmake_targets_map[package.name] = std::move(targets);
             }
-            else
+
+            if (has_targets_for_output)
             {
                 auto msg = msg::format(msgCMakeTargetsUsage, msg::package_name = bpgh.spec.name()).append_raw("\n\n");
                 msg.append_indent().append(msgCMakeTargetsUsageHeuristicMessage).append_raw('\n');
 
-                for (auto&& library_target_pair : library_targets)
+                for (const auto package_targets_pair : ret.cmake_targets_map)
```

Comment:
```suggestion
                for (auto&& package_targets_pair : ret.cmake_targets_map)
```

I believe auto&& as formerly written is more correct here: there's no reason to introduce an extra by-value copy.